SpatialGDK/ExternalSchemaCodegen/Programs/Improbable.CodeGen.Unreal/Serialization.cs

```diff
@@ -0,0 +1,443 @@
+        {
+            return $"{targetObjectName} = {Types.CollectionTypesToQualifiedTypes[Types.Collection.Option]}<{Types.GetQualifiedTypeName(optionType.InnerType, bundle)}>({GetValueTypeDeserialization(optionType.InnerType, schemaObjectName, fieldId)});";
+        }
+
+        private static string GetListTypeDeserialization(FieldDefinition.ListTypeRef listType, string schemaObjectName, string targetObjectName, string fieldName, string fieldId, Bundle bundle, bool wrapInBlock = false, bool targetIsOption = false)
+        {
+            var listInnerTypeName = Types.GetQualifiedTypeName(listType.InnerType, bundle);
+            var listName = $"{Text.SnakeCaseToPascalCase(fieldName)}List";
+            var builder = new StringBuilder();
+
+            switch (listType.InnerType.ValueTypeSelector)
+            {
+                case ValueType.Primitive:
+                    builder.AppendLine(GetPrimitiveListDeserialization(listType.InnerType.Primitive, schemaObjectName, targetObjectName, fieldName, fieldId, bundle, targetIsOption));
+                    break;
+                case ValueType.Type:
+                    builder.AppendLine($@"auto ListLength = Schema_GetObjectCount({schemaObjectName}, {fieldId});");
+                    builder.AppendLine($@"{Types.CollectionTypesToQualifiedTypes[Types.Collection.List]}<{listInnerTypeName}> {listName};");
+                    builder.AppendLine($@"{listName}.SetNum(ListLength);");
+                    builder.AppendLine($@"for (uint32 i = 0; i < ListLength; ++i)");
+                    builder.AppendLine($@"{{");
+                    builder.AppendLine(Text.Indent(1, $"{listName}[i] = {listInnerTypeName}::Deserialize(Schema_IndexObject({schemaObjectName}, {fieldId}, i));"));
+                    builder.AppendLine($@"}}");
+                    builder.AppendLine($@"{targetObjectName} = {listName};");
+                    break;
+                case ValueType.Enum:
+                    builder.AppendLine($@"auto ListLength = Schema_GetEnumCount({schemaObjectName}, {fieldId});");
+                    builder.AppendLine($@"{Types.CollectionTypesToQualifiedTypes[Types.Collection.List]}<{listInnerTypeName}> {listName};");
+                    builder.AppendLine($@"{listName}.SetNum(ListLength);");
+                    builder.AppendLine($@"for (uint32 i = 0; i < ListLength; ++i)");
+                    builder.AppendLine($@"{{");
+                    builder.AppendLine(Text.Indent(1, $"{listName}[i] = static_cast<{listInnerTypeName}>(Schema_IndexEnum({schemaObjectName}, {fieldId}, i));"));
+                    builder.AppendLine($@"}}");
+                    builder.AppendLine($@"{targetObjectName} = {listName};");
+                    break;
+                default:
+                    throw new InvalidOperationException("Trying to deserialize invalid ValueTypeReference");
+            }
+
+            if (wrapInBlock)
+            {
+                var blockWrapping = new StringBuilder();
```

Comment:
Just use a simple `$@""` and return that with everything interpolated into it - no need for a `StringBuilder`